pkg/logstructured/logstructured.go

```diff
@@ -294,28 +296,62 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 				}
 			}
 		}
+
+		if ctx.Err() == nil {
+			logrus.Error("ttl events watch channel was closed due to slow processing")
+		}
 	}()
 
 	return result
 }
 
 func (l *LogStructured) ttl(ctx context.Context) {
 	// vary naive TTL support
-	mutex := &sync.Mutex{}
-	for event := range l.ttlEvents(ctx) {
-		go func(event *server.Event) {
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Duration(event.KV.Lease) * time.Second):
-			}
-			mutex.Lock()
-			if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-				logrus.Errorf("failed to delete expired key: %v", err)
-			}
-			mutex.Unlock()
-		}(event)
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter())
+	go func() {
+		for l.handleTTLEvents(ctx, queue, 3) {
```

Comment:
Was 3 an arbitrary choice? Is there any reason we would want to skip deleting an expired item after a certain number of attempts?

Reply:
> Was 3 an arbitrary choice? Is there any reason we would want to skip deleting an expired item after a certain number of attempts?

Since the error information was not processed in the previous logic, three retries are temporarily added here. Maybe, we can not limit the number of times. If it fails, just put it back into the queue. What's your suggestion?

https://github.com/k3s-io/kine/blob/c5f2d892d70778fb9ad15208d4ca8617c3c794da/pkg/logstructured/logstructured.go#L302C1-L318C3

Reply:
If the error is that the revision is too old, it can be removed from the queue as the key has been updated. If there is some other error with the database or network, it should probably be requeued?

Reply:
I agree with your point of view, then I will remove the number limits.